### PR TITLE
[Driver] Clean up ownership of Action objects

### PR DESCRIPTION
**Pull request by jrose-apple**
Fixes a leak (rdar://problem/32518867) and just generally makes the world simpler now that we don't have to worry about ownership of these things. The Compilation will take care of them just like it does Jobs.

### CODE REVIEW

**Comment by jrose-apple** [comment]
@swift-ci Please asan test

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test

**Comment by gottesmm** [comment]
@jrose-apple I don't think the ASAN tests actually works, but I guess we will see. IIRC Xin looked into it last summer, but he was unable to figure out what was wrong (or something like that).

**Comment by gottesmm** [comment]
@shahmishal Is my memory correct?

**Comment by shahmishal** [comment]
it's working https://ci.swift.org/job/oss-swift-incremental-ASAN-RA-osx/

**Comment by gottesmm** [comment]
I meant for pull request testing.

**Comment by jrose-apple** [comment]
That failure…shouldn't have anything to do with this commit.

@swift-ci Please smoke test Linux

File: lib/Driver/Driver.cpp [hunk 1]
```diff
@@ -1803,14 +1764,10 @@ static StringRef getOutputFilename(Compilation &C,
...
   auto ShouldPreserveOnSignal = PreserveOnSignal::No;
 
-  // FIXME: Treat GeneratePCHJobAction as non-top-level (to get tempfile and not
-  // use the -o arg) even though, based on ownership considerations within the
-  // driver, it is stored as a "top level" JobAction.
   if (auto *PCHAct = dyn_cast<GeneratePCHJobAction>(JA)) {
     // For a persistent PCH we don't use an output, the frontend determines
     // the filename to use for the PCH.
     assert(!PCHAct->isPersistentPCH());
-    AtTopLevel = false;
```
Comment (graydon) [inline lib/Driver/Driver.cpp]: IIRC this was being done to provoke emitting a tempfile; is it still going to do so?

File: lib/Driver/Action.cpp [hunk 2]
```diff
@@ -18,12 +18,6 @@
 using namespace swift::driver;
 using namespace llvm::opt;
 
-JobAction::~JobAction() {
```
Comment (graydon) [inline lib/Driver/Action.cpp]: Hooray!

**Review by graydon — COMMENTED** [review]
Largely in favour of this! But I'm a little unclear on where the "generates a tempfile" vs. "doesn't" distinction that was also being represented in top-level actions has gone. I'm probably looking straight through it.

File: lib/Driver/Driver.cpp (hunk 1 above)
Comment (jrose-apple) [inline lib/Driver/Driver.cpp]: Hm, at one point I tried to remove the distinction between top-level actions and non-top-level actions and failed. I'll look into this again before merging.

Comment (jrose-apple) [inline lib/Driver/Driver.cpp]: Aha, I remember now: the PCH job is never top-level anymore, since it's not needed for ownership.

Comment (graydon) [inline lib/Driver/Driver.cpp]: Gotcha, ok. Cool, thanks!

**Review by graydon — APPROVED** [review]

**Pull request merged**